### PR TITLE
core: Add FactoryHashMap

### PR DESCRIPTION
**Pull request by AaronErhardt**
#### Summary

Supersedes #428 

@songww can you try this patch for your usecase?

#### Checklist

- [x] cargo fmt
- [x] cargo clippy
- [x] cargo test
- [x] updated CHANGES.md

### CODE REVIEW

File: relm4/src/factory/sync/collections/hashmap.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    }
+}
+
+/// A container similar to [`HashMap`] that can be used to store
+/// values of type [`FactoryComponent`].
+#[derive(Debug)]
+pub struct FactoryHashMap<K, C: FactoryComponent> {
+    widget: C::ParentWidget,
+    parent_sender: Sender<C::ParentInput>,
+    inner: HashMap<K, FactoryHandle<C>>,
```
Comment (songww) [inline relm4/src/factory/sync/collections/hashmap.rs:56]: I am being a bit demanding, but it would be really great if different hash types could be used.

**Review by songww — COMMENTED** [review]
Looks good to me.

File: relm4/src/factory/sync/collections/hashmap.rs (hunk 1 above)
Comment (AaronErhardt) [reply relm4/src/factory/sync/collections/hashmap.rs:56]: Can you elaborate what you mean with different hash types? I'm not sure if I understand what you mean.

File: examples/entry.rs [hunk 2]
```diff
@@ -27,6 +27,7 @@ impl FactoryComponent for Counter {
     type CommandOutput = ();
     type ParentInput = AppMsg;
     type ParentWidget = gtk::Box;
+    type Index = DynamicIndex;
```
Comment (euclio) [inline examples/entry.rs:30]: Should the macro default this to `DynamicIndex`?

**Review by euclio — APPROVED** [review]
Looks good. Not thrilled about adding yet another associated type to the trait but I'm not sure if there's a viable alternative.

File: relm4/src/factory/sync/collections/hashmap.rs (hunk 1 above)
Comment (AaronErhardt) [reply relm4/src/factory/sync/collections/hashmap.rs]: I think you meant different hashers, right? Anyway, I've just added support for that.

File: examples/entry.rs (hunk 2 above)
Comment (AaronErhardt) [reply examples/entry.rs]: Good suggestion. Is implemented now.

**Review by euclio — APPROVED** [review]

**Pull request merged**